### PR TITLE
AWS HSM wallet

### PR DESCRIPTION
**Pull request by AlexBHarley**
### Description

Integrates ContractKit with AWS HSMs. 

### Tested

Running an oracle client locally against an HSM I set up in AWS.

### CODE REVIEW

File: packages/contractkit/src/utils/ber-utils.ts
Comment (codyborn) [inline packages/contractkit/src/utils/ber-utils.ts]: It looks like we can just use the built in [.buffer property](https://nodejs.org/docs/latest/api/buffer.html#buffer_buf_buffer) in node versions >= 4.x
https://stackoverflow.com/a/12101012/677846

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (tkporter) [inline packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: wdyt of creating a new type for `{ v: number; r: Buffer; s: Buffer }`? It seems to be used a lot outside of this PR too

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (tkporter) [inline packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: `signTransaction` and `signPersonalMessage` here are pretty similar to the ones in the azure signer with the exception of the `sign` calls, do you think there's any way we could pull all that common code into `Signer` such that it will use the `sign` implementation that's specific to AWS/Azure?

**Review by tkporter — COMMENTED** [review]
Looks really really good
What are the test issues you're running into?

File: packages/contractkit/src/utils/ber-utils.ts
Comment (codyborn) [inline packages/contractkit/src/utils/ber-utils.ts]: I'm unable to find info on this `valueBlockType` constructor. Is it guaranteed to have a value array property with at least two values?

File: packages/contractkit/src/wallets/aws-hsm-wallet.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/aws-hsm-wallet.ts]: It seems like this value should be configurable via environment vars so that we can support a variety of regions.

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: Can we re-use `recoverKeyIndex` from [here](https://github.com/celo-org/celo-monorepo/blob/11814bdb3b44bddb3c19a9ec0f439ad45e86b6bc/packages/contractkit/src/utils/azure-key-vault-client.ts#L147)? From what I read it's possible for this key to be in the set [0,1,2,3], although 2,3 are very rare:
https://ethereum.stackexchange.com/questions/42455/during-ecdsa-signing-how-do-i-generate-the-recovery-id

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: Is this signing operation non-deterministic? What causes it to become canonical?

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (codyborn) [reply packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: +1 There's one here, but it should probably be moved to a common signature lib:
https://github.com/celo-org/celo-monorepo/blob/11814bdb3b44bddb3c19a9ec0f439ad45e86b6bc/packages/contractkit/src/utils/azure-key-vault-client.ts#L213

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (codyborn) [reply packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: I think that's a good idea.

File: packages/contractkit/src/wallets/remote-wallet.test.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/remote-wallet.test.ts]:
```suggestion
describe('RemoteWallet class', () => {
```

File: packages/contractkit/src/wallets/remote-wallet.test.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/remote-wallet.test.ts:8]: Thanks for adding this!!

File: packages/contractkit/src/wallets/aws-hsm-wallet.test.ts
Comment (codyborn) [inline packages/contractkit/src/wallets/aws-hsm-wallet.test.ts:20]: I believe you'll have to add this line to read from env file:
`require('dotenv').config()`

**Review by codyborn — CHANGES_REQUESTED** [review]
Looks good! Before you check-in please confirm that you've ran the tests using the non-mock versions of AWS and Azure.

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (AlexBHarley) [reply packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: I've really struggled getting this to work... The ecdsaRecover fn I an using just calls into the `secp256k1` but I can't figure out what is going wrong.

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (AlexBHarley) [reply packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: I believe signing is non-deterministic yes.

File: packages/contractkit/src/wallets/signers/aws-hsm-signer.ts
Comment (AlexBHarley) [reply packages/contractkit/src/wallets/signers/aws-hsm-signer.ts]: This is a bit blocked on getting the `recoverKeyIndex` function to work... I have struggled with the implementation that we have in this repo already.

File: packages/contractkit/src/utils/ber-utils.ts
Comment (AlexBHarley) [reply packages/contractkit/src/utils/ber-utils.ts]: Added a check just in case

File: packages/contractkit/src/utils/azure-key-vault-client.ts
Comment (codyborn) [inline packages/contractkit/src/utils/azure-key-vault-client.ts:70]: This line was changed to fix the `value must be of type Uint8Array` issue. I believe some change in monorepo to our dependencies caused this to start failing.

File: packages/contractkit/src/utils/signing-utils.ts
Comment (codyborn) [inline packages/contractkit/src/utils/signing-utils.ts:250]: @tkporter I tried consolidating this to use the util `publicKeyToAddress` method; however it introduced more problems. `ethUtil.pubToAddress` method does additional cleansing that allows it to succeed under more conditions.

File: packages/contractkit/src/utils/signature-utils.ts
Comment (codyborn) [inline packages/contractkit/src/utils/signature-utils.ts:14]: New method to move more common logic between AWS + Azure into signing-utils

**Review by codyborn — APPROVED** [review]
Approving changes, will wait for @tkporter to sign off before merging.

File: packages/contractkit/src/utils/ber-utils.ts
Comment (tkporter) [inline packages/contractkit/src/utils/ber-utils.ts:15]: just want to make sure it's safe to assume `values.length >= 2`

File: packages/contractkit/src/utils/ber-utils.ts
Comment (tkporter) [inline packages/contractkit/src/utils/ber-utils.ts:9]: are these functions pretty generic or specific to expected responses from Azure/AWS?

File: packages/contractkit/src/utils/signing-utils.ts
Comment (tkporter) [inline packages/contractkit/src/utils/signing-utils.ts]: not sure if `console.log` is used in contractkit, I think instead `debug` from line 14 is used

File: packages/contractkit/src/utils/azure-key-vault-client.ts
Comment (tkporter) [inline packages/contractkit/src/utils/azure-key-vault-client.ts]: should we use `thirtyTwo` and `sixtyFour` here?

**Review by tkporter — APPROVED** [review]
Looks great! some very small things

File: packages/contractkit/src/utils/ber-utils.ts
Comment (codyborn) [reply packages/contractkit/src/utils/ber-utils.ts:15]: Good point. Either way we'll fail, but it's better to fail with a more descriptive error.

File: packages/contractkit/src/utils/signing-utils.ts
Comment (codyborn) [reply packages/contractkit/src/utils/signing-utils.ts]: good catch

File: packages/contractkit/src/utils/ber-utils.ts
Comment (codyborn) [reply packages/contractkit/src/utils/ber-utils.ts:9]: This method takes a BER encoded key with ASN1 serialization protocol and converts it to a BigNumber format, which our libraries are using as the standard. AWS sends the key in this format so although this method is generic, I don't expect other libraries to use this.

**Pull request merged**
